concrete-core-bench/src/benchmark.rs

```diff
@@ -0,0 +1,144 @@
+use criterion::measurement::WallTime;
+use criterion::{BenchmarkGroup, BenchmarkId, Criterion};
+use std::cmp::min;
+use std::ops::Add;
+use std::time::{Duration, Instant};
+
+/// Default value for the batch size.
+pub const DEFAULT_BATCH_SIZE: u64 = 1000;
+
+/// An extension to the `Fixture` trait for benchmarking with criterion.
+pub trait BenchmarkFixture<Precision: IntegerPrecision, Engine: AbstractEngine, RelatedEntities>:
+    Fixture<Precision, Engine, RelatedEntities> + Sized
+{
+    /// Benchmarks all the parameters defined by the fixture.
+    fn bench_all_parameters(
+        maker: &mut Maker,
+        engine: &mut Engine,
+        criterion: &mut Criterion,
+        batch_size: Option<u64>,
+    ) {
+        let mut group = criterion.benchmark_group(format!(
+            "{}<{}, {}, {}>",
+            type_name::<Self>(),
+            type_name::<Precision>(),
+            type_name::<Engine>(),
+            type_name::<RelatedEntities>()
+        ));
+        for params in Self::generate_parameters_iterator() {
+            Self::bench(maker, engine, params, &mut group, batch_size);
+        }
+        group.finish();
+    }
+
+    /// Benchmark one parameter in the fixture benchmark group.
+    fn bench(
+        maker: &mut Maker,
+        engine: &mut Engine,
+        parameters: Self::Parameters,
+        fixture_group: &mut BenchmarkGroup<WallTime>,
+        batch_size: Option<u64>,
+    ) {
+        let batch_size = batch_size.unwrap_or(DEFAULT_BATCH_SIZE);
+
+        // We generate the prototypes once (used for all repetitions of the benchmark).
+        let repetition_proto = Self::generate_random_repetition_prototypes(&parameters, maker);
+        let sample_proto =
+            Self::generate_random_sample_prototypes(&parameters, maker, &repetition_proto);
+
+        // The benchmark itself does not use the standard iterators of `criterion`, because the
+        // pre-context is passed by value, and the post-context must be gathered to be
+        // properly disposed afterward.
+        fixture_group.bench_with_input(
+            BenchmarkId::from_parameter(format!("{:?}", parameters)),
+            &(repetition_proto, sample_proto),
+            |b, (repetition_proto, sample_proto)| {
+                b.iter_custom(|iterations_to_go| {
+                    // We initialize the vectors which will store the input and output values of the
+                    // batch.
+                    let mut inputs_vec: Vec<Self::PreExecutionContext> =
+                        Vec::with_capacity(batch_size as usize);
+                    let mut outputs_vec: Vec<Self::PostExecutionContext> =
+                        Vec::with_capacity(batch_size as usize);
+
+                    // We iterate until we reached the number of iterations to be performed.
+                    let mut iterations_counter = 0;
+                    let mut duration = Duration::new(0, 0);
+                    while iterations_counter < iterations_to_go {
```

Comment:
is it possible to run this with a range starting at 0, finishing at iteration_to_go and with a stride of batch_size?

Avoids potential errors when handling loop logic (in python it's done all the time, not sure for Rust)

Reply:
ok no problem good to know !